### ISSUE TITLE
Switch to git flow branching model and standardize commit message

### ISSUE THREAD

**Issue by wayneyu** [wayneyu/merapp#35]
What do you guys think of using a standardized workflow for git version control:
Article explaining the concepts:
http://nvie.com/posts/a-successful-git-branching-model/
A tool to do git-flow:
http://danielkummer.github.io/git-flow-cheatsheet/

Basically, the idea is to have a feature branch for every issue we have. The person that picks up an issue starts his work by branching off develop. Then he will commit his changes locally and push to a remote feature branch and once the changes are ready, he will merge them into develop and push develop up.  This way, we can help writing and reviewing each other's code without touching develop which will make the develop branch cleaner and easier to see what major works have been done.

Also, we should have a standard form for the commit message:
I suggest using "Issue #IssueNumber: Commit message"
This way all the commits will be automatically linked to the issue tracker which will facilitate code tracking and discussions.

Let me know what you think.

**Comment by BernhardKonrad** [comment]
Great idea! I've seen this page before, and had always wanted to use it in a project.

I think a model like this will be necessary when we distinguish between `master` and `develop`, that is, when we need a version that is always stable. So I suggest we make the branching model a suggestion, but not a requirement at this point.

**Comment by wayneyu** [comment]
Yes, we don't have to follow the model exactly right now.

I suggest we create at least a separate branch for each issue and include the issue number in each commit message instead of all of us commit to the same branch for every issue.

By keeping the smaller commits made towards resolving each issue in a separate branch, we can track what's been done and what else needs to be done for the issue and more importantly allow anyone to pick up where the issue was last left off and start making changes right away.

**Comment by BernhardKonrad** [comment]
Okay, one branch per issue sounds good. Branch name is then `issue123`?

**Comment by wayneyu** [comment]
Sure. And we can use "Issue #123: Message" for commit messages.

**Comment by BernhardKonrad** [comment]
Great. When merging on the command line, please make sure to use [`--no-ff`](http://stackoverflow.com/questions/6701292/git-fast-forward-vs-no-fast-forward-merge), [GitHub's merge button](https://github.com/blog/843-the-merge-button) does this by default.

**Comment by BernhardKonrad** [comment]
What should we do for little changes that are not an `issue`? Create one and then fix it on the new, corresponding branch?

**Comment by BernhardKonrad** [comment]
Also, what should we do with `master`? Delete it for now?

**Comment by wayneyu** [comment]
Yes, we should create a new issue, so that every commit can be tracked by an issue.

We should never delete `master`, as it holds the production releases. Same applies to `develop`.

**Comment by wayneyu** [comment]
@WillThompson , we started to have a standardized workflow for branching and commiting a couple of days ago.
Basically the steps are:
1) Assign yourself or get assigned to an issue in Github's issue tracker.
2) Checkout and branch off develop:

```
git checkout develop;
git pull
git branch issue-#number
git checkout issue-#number
// make changes 
git add .
git commit -m "Issue #number: Commit message"
git push -u origin issue-#number
```

3) If you want others to review your code, you can create a pull request in GitHub, otherwise

```
git checkout develop
git pull
git merge issue-#number
// resolve any conflicts
git push
// optional: push to heroku
```

**Comment by BernhardKonrad** [comment]
If that's confusing at this point, a first step would be to use the format `git commit -m "Issue #number: Commit message" in each commit message, and open/close GitHub issues as you start/finish work on stuff. This makes the history a little more sane with several people working on it.

**Comment by BernhardKonrad** [comment]
@wayneyu What do you think:
Let's start using `master`, and use `master` exclusively for pushing to Heroku for deployment in regular (weekly?) intervals. As in releases. `develop` will hold the most advanced version, and we merge `develop` into `master` whenever we push to Heroku. Individual issues are still developed on entirely separate `issue-...` branches.

**Comment by BernhardKonrad** [comment]
@wayneyu Could you please add Will to heroku?
